dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -0,0 +1,163 @@
+    thread_pool_.AddWork([&curr_batch, &curr_tensor_list, data_idx](int tid) {
+        curr_batch[data_idx]->read_sample_f(curr_tensor_list.raw_mutable_data(),
+                                            curr_tensor_list.tensor_offset(data_idx) *
+                                            curr_tensor_list.type().size(),
+                                            curr_tensor_list.nbytes());
+      });
+  }
+  thread_pool_.RunAll();
+}
+
+void PermuteHelper(const TensorShape<> &plain_shapes, std::vector<int64_t> &perm_shape,
+                  std::vector<int> &perm) {
+  int n_dims = plain_shapes.size();
+  if (perm.empty()) {
+    perm.resize(n_dims);
+    for (int i = 0; i < n_dims; ++i) {
+      perm[i] = n_dims - i - 1;
+    }
+  }
+  for (int i = 0; i < n_dims; ++i) {
+    perm_shape[i] = plain_shapes[perm[i]];
+  }
+}
+
+void NumpyReaderGPU::RunImpl(DeviceWorkspace &ws) {
+  TensorListShape<> shape(batch_size_);
+  // use vector for temporarily storing shapes
+  std::vector<TensorShape<>> tmp_shapes;
+  std::vector<TensorShape<>> transpose_shapes;
+  std::vector<int> perm;
+  std::vector<int64_t> perm_shape;
+
+  perm.reserve(GetSampleShape(0).size());
+  perm_shape.resize(GetSampleShape(0).size());
+
+  for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
+    const auto& imfile = GetSample(sample_idx);
+    auto plain_shape = GetSampleShape(sample_idx);
+    if (imfile.meta == "transpose:false") {
+      tmp_shapes.push_back(plain_shape);
+    } else {
+      PermuteHelper(plain_shape, perm_shape, perm);
+      tmp_shapes.push_back(perm_shape);
+      transpose_shapes.push_back(plain_shape);
+    }
+  }
+  auto ref_type = GetSampleType(0);
+  shape = TensorListShape<>(tmp_shapes);
+  ws.Output<GPUBackend>(0).Resize(shape, ref_type);
+
+  auto &image_output = ws.Output<GPUBackend>(0);
+
+  SmallVector<int64_t, 256> copy_sizes;
+  copy_sizes.reserve(batch_size_);
+  SmallVector<const void *, 256> copy_from;
+  copy_from.reserve(batch_size_);
+  SmallVector<void *, 256> copy_to;
+  copy_to.reserve(batch_size_);
+
+  SmallVector<const void *, 256> transpose_from;
+  transpose_from.reserve(batch_size_);
+  SmallVector<void *, 256> transpose_to;
+  transpose_to.reserve(batch_size_);
+
+
+  for (int data_idx = 0; data_idx < batch_size_; ++data_idx) {
+    const auto& imfile = GetSample(data_idx);
+    if (imfile.meta == "transpose:false") {
+      copy_from.push_back(GetSampleRawData(data_idx));
+      copy_to.push_back(image_output.raw_mutable_tensor(data_idx));
+      copy_sizes.push_back(shape.tensor_size(data_idx));
+    } else {
+      transpose_from.push_back(GetSampleRawData(data_idx));
+      transpose_to.push_back(image_output.raw_mutable_tensor(data_idx));
+    }
+    image_output.SetSourceInfo(data_idx, imfile.image.GetSourceInfo());
+  }
+
+  // use copy kernel for plan samples
+  if (!copy_sizes.empty()) {
+    ref_type.template Copy<GPUBackend, GPUBackend>(copy_to.data(), copy_from.data(),
```

Comment:
I guess detecting if all samples don't require transpose and reading directly to a target batch (that we already know size of, as we parsed it), would be a potential optimization. Or maybe doing it at least as one big D2D copy?

Reply:
>I guess detecting if all samples don't require transpose and reading directly to a target batch (that we already know size of, as we parsed it), would be a potential optimization. 

It is not possible as in the prefetch we don't know the operator output buffer, RunImpl knows that but the prefetched batch is already there. It is not contiguous so we need that copy (we cannot simply swap buffers).

As you see only on copy kernel is invoked here. I don't think we can do anything faster to copy from non-contiguous tensor vector into tensor list.

Reply:
Correction - I forgot we have prefetched_batch_tensors_ that is contiguous. We can swap it if no transposition is needed at all.
Still ad. 1 - it is crated asynchronously in the prefetch thread so nothing beyond that can be done.